blog/_posts/2018-06-19.md

```diff
@@ -0,0 +1,496 @@
+
+The second improvement consists in using a compact memory layout for `Array` object
+whose element type is a `Union` of bits types, i.e. immutable types which contain
+no references (see the [`isbits`](https://docs.julialang.org/en/latest/base/base/#Base.isbits)
+function). This includes `Missing` and basic types such as `Int`, `Float64`,
+`Complex{Float64}` and `Date`. When `T` is a bits type, `Array{Union{Missing,T}}`
+objects are internally represented as a pair of arrays of the same size:
+an `Array{T}` holding non-missing values and uninitialized memory for missing values;
+and an `Array{UInt8}` storing a *type tag* indicating whether each entry
+is of type `Missing` or `T`.
+
+This layout consumes slightly more memory than the sentinel approach, as the
+type tag part occupies one byte for each entry. But this overhead is reasonable:
+for example, the memory usage of an `Array{Union{Missing,Float64}}` is only
+12.5% higher than that of an `Array{Float64}`. Compared with the sentinel approach,
+it has the advantage of being fully generic (as we have seen above).
+Actually, this mechanism can be used in other situations, for example with `Union{Nothing,Int}`
+(which is the element type of the array returned by `indexin` in Julia 0.7).
+
+Arrays of non-bits types with missing values were already and continue to be represented
+as efficiently as their counterparts without missing values. Indeed, such arrays consist of
+pointers to the actual objects which live in a different memory area. Missing values
+can be represented as a special pointer just like non-missing values. This is notably
+the case for `Array{Union{Missing,String}}`.
+
+The efficient memory layout of `Array` in the presence of missing values makes it
+unnecessary to use dedicated array types like
+[`DataArray`](https://github.com/JuliaStats/DataArrays.jl). In fact, the layout
+of the `DataArray` type is very similar to that of `Array{Union{Missing,T}}`
+described above. The only difference is that it uses a `BitArray` rather than an
+`Array{UInt8}` to indicate whether a value is missing, therefore taking 1 bit per
+entry rather than 8 bits. Even if it consumes more memory, the `Array{UInt8}` mask
+approach is faster (at least in the current state of `BitArray`), and it generalizes
+to `Union`s of more than two types. However, we are aware that other implementations
+such as [PostgreSQL](https://www.postgresql.org/docs/9.5/static/storage-page-layout.html)
+or [Apache Arrow](https://arrow.apache.org/docs/memory_layout.html#null-bitmaps)
+use bitmaps equivalent to `BitArray`.
+
+This efficient representation allows the compiler to generate very efficient code when
+processing arrays with missing values. For example, the following function is one of
+the most straightforward ways of computing the sum of non-missing values in an array:
+
+    function sum_nonmissing(X::AbstractArray)
+        s = zero(eltype(X))
+        @inbounds @simd for x in X
+            if x !== missing
```

Comment:
`!ismissing`?

Reply:
Read below... :-p

(Currently it's much slower unfortunately.)

Reply:
Whoops. :sweat_smile: